ban/commands/ignsna.py

```diff
@@ -0,0 +1,111 @@
+import glob
+import os
+
+from peewee import IntegrityError
+
+from ban.commands import command, report
+from ban.core.models import (HouseNumber, Locality, Municipality, Position,
+                             Street, PostCode)
+from .helpers import iter_file, session, batch
+
+__namespace__ = 'import'
+
+
+@command
+def ignsna(path, **kwargs):
+    """Import from IGN/Laposte BDUNI
+    :param path: directory location of the files hsp7xxxx.ai, hsv7xxxx.ai and hsw4xxxx.ai"""
+
+    municipality_zipcode_file = glob.glob(os.path.join(path, 'hsp7*.ai'))
+    # street_file = glob.glob(os.path.join(path, 'hsv7*.ai'))
+    # number_file = glob.glob(os.path.join(path, 'hsw4*.ai'))
+
+    if municipality_zipcode_file is not None:
+        max_value = get_max_line(municipality_zipcode_file[0])
+        lines = list(get_lines(municipality_zipcode_file[0]))
+        # ToDo: test the performance of single threaded and multi-threaded processes
+        batch(process_municipality_file, lines, max_value=max_value)
+
+        # ToDo: uncomment and complete when process needed
+        # if street_file is not None:
+        #     process_streetFile(street_file[0])
+        #
+        # if number_file is not None:
+        #     process_numberFile(number_file[0])
+
+
+@session
+def process_municipality_file(line):
+    # line = lines[x]
+    if line[50] == 'M':
+        insee = line[6:11]
+        # name = line[11:49]
+        # name = name.strip()
+        zip_code = line[89:94]
+        # old_insee = line[126:131]
+        # old_insee = old_insee.strip()
+
+        zip_code_bean = PostCode.create_or_get(code=zip_code, version='1')
+        try:
+            municipality = Municipality.get(Municipality.insee == insee)
+            code = municipality.postcodes
+            if not (zip_code_bean[0]) in code:
+                if zip_code_bean[0]:
+                    try:
+                        municipality.postcodes.add(zip_code_bean[0])
+                    except IntegrityError:
```

Comment:
And this is it :) There can be only one relation between municipality and postal code.
If we add an already existing postal code to a municipality or conversely an existing municipality to a postal code, the integrity error arises.
I'm talking to the import consecutively (to err is human) and not at the same time, it will not happen (I hope). :)